File: lib/device_app/localSaver/custom_app_model.dart
import 'dart:typed_data';

class CustomAppModel {
  final String packageName;
  final String displayName;
  final Uint8List? icon;
  final int isSystemApp;
  final int isFavorite;
  final int countdown;
  final int lastOpened;
  final int todayLimit;
  final int todayUsage;
  final int extraLimit;
  final int sessionLimit;
  final int sessionUsage;

  CustomAppModel({
    required this.packageName,
    required this.displayName,
    this.icon,
    required this.isSystemApp,
    required this.isFavorite,
    required this.countdown,
    required this.lastOpened,
    required this.todayLimit,
    required this.todayUsage,
    required this.extraLimit,
    this.sessionLimit = 0,
    this.sessionUsage = 0,
  });

  factory CustomAppModel.fromMap(Map<String, dynamic> map) {
    dynamic iconData = map['icon'];
    Uint8List? iconBytes;
    if (iconData != null) {
      if (iconData is Uint8List) {
        iconBytes = iconData;
      } else if (iconData is List) {
        iconBytes = Uint8List.fromList(List<int>.from(iconData));
      }
    }

    return CustomAppModel(
      packageName: map['packageName'] as String? ?? '',
      displayName: map['displayName'] as String? ?? '',
      icon: iconBytes,
      isSystemApp: map['isSystemApp'] as int? ?? 0,
      isFavorite: map['isFavorite'] as int? ?? 0,
      countdown: map['countdown'] as int? ?? 0,
      lastOpened: map['lastOpened'] as int? ?? 0,
      todayLimit: map['todayLimit'] as int? ?? 0,
      todayUsage: map['todayUsage'] as int? ?? 0,
      extraLimit: map['extraLimit'] as int? ?? 0,
      sessionLimit: map['sessionLimit'] as int? ?? 0,
      sessionUsage: map['sessionUsage'] as int? ?? 0,
    );
  }

  @override
  String toString() {
    return 'CustomAppModel(packageName: $packageName, displayName: $displayName, hasIcon: ${icon != null}, isSystemApp: $isSystemApp, isFavorite: $isFavorite, countdown: $countdown, lastOpened: $lastOpened, todayLimit: $todayLimit, todayUsage: $todayUsage, extraLimit: $extraLimit, sessionLimit: $sessionLimit, sessionUsage: $sessionUsage)';
  }
}

File: android/app/src/main/kotlin/com/example/testproject/CustomAppModel.kt
package com.example.testproject

/**
 * CustomAppModel:
 * Yeh model class Dart side ke CustomAppModel.dart ko mirror karti hai.
 * Har ek app jiska limit ya usage hum track kar rahe hain, uska data is object me store hoga.
 */
data class CustomAppModel(
    // App ka package name (Unique ID) jaise: "com.whatsapp"
    val packageName: String,

    // App ka visible display name (Label) jaise: "WhatsApp"
    val displayName: String,

    // App ka icon blob (binary image data) jo SQLite database me saved hai
    val icon: ByteArray? = null,

    // Kya ye system app hai? (1 = System App, 0 = Installed/User App)
    val isSystemApp: Int,

    // Kya user ne is app ko favorited list me add kiya hai? (1 = Yes, 0 = No)
    val isFavorite: Int,

    // App open hone se pehle user ke liye screen pause delay (Mindful Delay in seconds)
    val countdown: Int,

    // App last time kis timestamp par open ki gayi thi
    val lastOpened: Long,

    // Aaj ke din ka user-set daily limit (Milliseconds me)
    var todayLimit: Long = 0,

    // Aaj ke din ka user ka accumulated usage time (Milliseconds me)
    var todayUsage: Long = 0,

    // Extra allowed extended limit time (2 minutes extend window button click par - Milliseconds me)
    var extraLimit: Long = 0,

    // Temporary active session limit (In-memory, jaise 5 min or 10 min - Milliseconds me)
    var sessionLimit: Long = 0,

    // Current temporary active session usage (In-memory - Milliseconds me)
    var sessionUsage: Long = 0
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as CustomAppModel

        if (packageName != other.packageName) return false
        if (displayName != other.displayName) return false
        if (icon != null) {
            if (other.icon == null) return false
            if (!icon.contentEquals(other.icon)) return false
        } else if (other.icon != null) return false
        if (isSystemApp != other.isSystemApp) return false
        if (isFavorite != other.isFavorite) return false
        if (countdown != other.countdown) return false
        if (lastOpened != other.lastOpened) return false
        if (todayLimit != other.todayLimit) return false
        if (todayUsage != other.todayUsage) return false
        if (extraLimit != other.extraLimit) return false
        if (sessionLimit != other.sessionLimit) return false
        if (sessionUsage != other.sessionUsage) return false

        return true
    }

    override fun hashCode(): Int {
        var result = packageName.hashCode()
        result = 31 * result + displayName.hashCode()
        result = 31 * result + (icon?.contentHashCode() ?: 0)
        result = 31 * result + isSystemApp
        result = 31 * result + isFavorite
        result = 31 * result + countdown
        result = 31 * result + lastOpened.hashCode()
        result = 31 * result + todayLimit.hashCode()
        result = 31 * result + todayUsage.hashCode()
        result = 31 * result + extraLimit.hashCode()
        result = 31 * result + sessionLimit.hashCode()
        result = 31 * result + sessionUsage.hashCode()
        return result
    }
}
